File: delta/scripts/generate-icons.mjs
/**
 * Generates PWA icons (192x192 and 512x512) as PNG files
 * using only Node built-ins + a tiny inline SVG → PNG via canvas API.
 *
 * Run: node scripts/generate-icons.mjs
 *
 * Requires no extra npm packages — uses the @resvg/resvg-js approach
 * via the built-in sharp if available, otherwise falls back to a
 * minimal base64-encoded PNG.
 */

import { writeFileSync, mkdirSync } from "fs";
import { createCanvas } from "canvas";         // optional — only if available
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const outDir = path.join(__dirname, "../public/icons");
mkdirSync(outDir, { recursive: true });

function generateIcon(size) {
  const canvas = createCanvas(size, size);
  const ctx = canvas.getContext("2d");

  // Background — violet gradient approximated as solid
  const radius = size * 0.18;
  ctx.beginPath();
  ctx.moveTo(radius, 0);
  ctx.arcTo(size, 0, size, size, radius);
  ctx.arcTo(size, size, 0, size, radius);
  ctx.arcTo(0, size, 0, 0, radius);
  ctx.arcTo(0, 0, size, 0, radius);
  ctx.closePath();

  const grad = ctx.createLinearGradient(0, 0, size, size);
  grad.addColorStop(0, "#6d28d9");
  grad.addColorStop(1, "#4f46e5");
  ctx.fillStyle = grad;
  ctx.fill();

  // Zap / lightning bolt
  const s = size;
  ctx.strokeStyle = "white";
  ctx.lineWidth = s * 0.085;
  ctx.lineCap = "round";
  ctx.lineJoin = "round";
  ctx.fillStyle = "white";

  // Scale the polygon points to current size (original viewBox 24x24)
  const scale = s / 24;
  const points = [
    [13, 2], [3, 14], [12, 14], [11, 22], [21, 10], [12, 10], [13, 2],
  ];

  ctx.beginPath();
  points.forEach(([x, y], i) => {
    const px = x * scale;
    const py = y * scale;
    i === 0 ? ctx.moveTo(px, py) : ctx.lineTo(px, py);
  });
  ctx.closePath();
  ctx.fill();

  return canvas.toBuffer("image/png");
}

try {
  writeFileSync(path.join(outDir, "icon-192.png"), generateIcon(192));
  console.log("✅ icon-192.png generated");
  writeFileSync(path.join(outDir, "icon-512.png"), generateIcon(512));
  console.log("✅ icon-512.png generated");
} catch (e) {
  console.error("canvas not available, using fallback SVG-as-PNG placeholder");
  // Fallback: write a minimal 1x1 transparent PNG and log a warning
  const placeholder = Buffer.from(
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==",
    "base64"
  );
  writeFileSync(path.join(outDir, "icon-192.png"), placeholder);
  writeFileSync(path.join(outDir, "icon-512.png"), placeholder);
  console.warn("⚠️  Placeholder icons written. Run: npm install canvas && node scripts/generate-icons.mjs");
}

File: delta/scripts/generate-icons.py
"""
Generates PWA icons (192x192 and 512x512) as PNG files.
Uses only Python stdlib — no Pillow or other packages needed.
Run: python3 scripts/generate-icons.py
"""

import struct
import zlib
import os
import math

OUT_DIR = os.path.join(os.path.dirname(__file__), "../public/icons")
os.makedirs(OUT_DIR, exist_ok=True)


# ── Minimal PNG writer ────────────────────────────────────────────────────────

def png_chunk(name: bytes, data: bytes) -> bytes:
    c = zlib.crc32(name + data) & 0xFFFFFFFF
    return struct.pack(">I", len(data)) + name + data + struct.pack(">I", c)

def write_png(path: str, pixels: list[list[tuple[int,int,int,int]]], size: int):
    sig = b"\x89PNG\r\n\x1a\n"
    ihdr = png_chunk(b"IHDR", struct.pack(">IIBBBBB", size, size, 8, 2, 0, 0, 0))

    raw = bytearray()
    for row in pixels:
        raw.append(0)  # filter type None
        for r, g, b, _ in row:
            raw.extend([r, g, b])

    idat = png_chunk(b"IDAT", zlib.compress(bytes(raw), 9))
    iend = png_chunk(b"IEND", b"")

    with open(path, "wb") as f:
        f.write(sig + ihdr + idat + iend)


# ── Drawing helpers ───────────────────────────────────────────────────────────

def lerp(a, b, t):
    return a + (b - a) * t

def gradient_bg(x, y, size):
    """Violet diagonal gradient: #6d28d9 → #4f46e5"""
    t = (x + y) / (2 * size)
    r = int(lerp(0x6d, 0x4f, t))
    g = int(lerp(0x28, 0x46, t))
    b = int(lerp(0xd9, 0xe5, t))
    return (r, g, b, 255)

def rounded_rect_mask(x, y, size, radius):
    """Returns True if pixel (x,y) is inside a rounded rectangle."""
    rx, ry = radius, radius
    cx = max(rx, min(x, size - 1 - rx))
    cy = max(ry, min(y, size - 1 - ry))
    dx, dy = x - cx, y - cy
    return (dx * dx + dy * dy) <= rx * rx

def point_in_polygon(px, py, poly):
    """Ray-casting inside-polygon test."""
    n = len(poly)
    inside = False
    j = n - 1
    for i in range(n):
        xi, yi = poly[i]
        xj, yj = poly[j]
        if ((yi > py) != (yj > py)) and (px < (xj - xi) * (py - yi) / (yj - yi) + xi):
            inside = not inside
        j = i
    return inside

def zap_polygon(size):
    """
    Returns the lightning-bolt polygon scaled to the icon size.
    Original viewBox: 24x24
    Points: 13 2 | 3 14 | 12 14 | 11 22 | 21 10 | 12 10 | 13 2
    Padded 20% on each side.
    """
    pad = size * 0.18
    draw_size = size - 2 * pad
    scale = draw_size / 24

    raw = [(13,2),(3,14),(12,14),(11,22),(21,10),(12,10)]
    return [(x * scale + pad, y * scale + pad) for x, y in raw]


def generate_icon(size: int) -> list[list[tuple[int,int,int,int]]]:
    radius = int(size * 0.18)
    poly   = zap_polygon(size)
    pixels = []

    for y in range(size):
        row = []
        for x in range(size):
            # 1. Outside rounded rect → transparent
            if not rounded_rect_mask(x, y, size, radius):
                row.append((0, 0, 0, 0))
                continue

            # 2. Inside lightning bolt → white
            if point_in_polygon(x + 0.5, y + 0.5, poly):
                row.append((255, 255, 255, 255))
                continue

            # 3. Background gradient
            row.append(gradient_bg(x, y, size))

        pixels.append(row)

    return pixels


# ── Generate ──────────────────────────────────────────────────────────────────

for sz in [192, 512]:
    path = os.path.join(OUT_DIR, f"icon-{sz}.png")
    print(f"Generating {sz}x{sz}...", end=" ", flush=True)
    pixels = generate_icon(sz)
    write_png(path, pixels, sz)
    print(f"✅  saved → {path}")

print("\nAll PWA icons generated successfully.")
